torch/csrc/jit/passes/onnx/constant_fold.cpp

```diff
@@ -0,0 +1,187 @@
+    }
+    updated_val = inputTensorValues[0];
+    for (size_t i = 0; i < axesAttr.size(); ++i) {
+      updated_val = at::narrow(updated_val, axesAttr[i], startsAttr[i], endsAttr[i] - startsAttr[i]);
+    }
+  }  
+  else if (node->kind() == onnx::Concat) {
+    updated_val = at::cat(at::TensorList(inputTensorValues), node->i(attr::axis));
+  }
+  else if (node->kind() == onnx::Unsqueeze) {
+    assert(inputTensorValues.size() == 1);
+    if (!node->hasAttributeS("axes")) {
+      throw std::runtime_error("Missing attribute 'axes' in onnx::Unsqueeze op.");
+    }
+    updated_val = inputTensorValues[0];
+    for (auto axis: node->is(attr::axes)) {
+      updated_val = at::unsqueeze(updated_val, axis);
+    }
+  }
+  else if (node->kind() == onnx::Transpose) {
+    assert(inputTensorValues.size() == 1);
+    if (!node->hasAttributeS("perm")) {
+      throw std::runtime_error("Missing attribute 'perm' in onnx::Transpose op.");
+    }
+    updated_val = inputTensorValues[0].permute(node->is(attr::perm));
+  }
+  else {
+    updated_val = at::empty({0});
+    auto qe = updated_val.size(0);
+  }
+  return updated_val;
+}
+
+static bool isConstant(Value* val, const ValueToParamPairMap& valsToParamsMap) {
+  auto parentNode = val->node();
+  return (parentNode->kind() == prim::Param && 
+          valsToParamsMap.find(val) != valsToParamsMap.end()) || // Checks val is a parameter and not a real input
+         (parentNode->kind() == onnx::Constant && !parentNode->mustBeNone() &&
+          parentNode->kindOf(attr::value) == AttributeKind::t); // Check other types?
+}
+
+static std::vector<at::Tensor> getValues(Node* node, const ValueToParamPairMap& valsToParamsMap) {
+  size_t numInputs = node->inputs().size();
+  std::vector<at::Tensor> inputTensorValues;
+  inputTensorValues.reserve(numInputs);
+  for (auto val : node->inputs()) {
+    if (val->node()->kind() == prim::Param) {
+      auto itr = valsToParamsMap.find(val);
+      if(itr == valsToParamsMap.end()) {
+        throw std::runtime_error("getValues: Input value not found amongst constant parameters.");
+      }
+      inputTensorValues.push_back(itr->second.second);
+    }
+    else if (val->node()->kind() == onnx::Constant) {
+      inputTensorValues.push_back(val->node()->t(attr::value));
+    }
+    else {
+      throw std::runtime_error("getValues: Unsupported kind of constant node found.");
+    }
+  }
+  AT_ASSERT(inputTensorValues.size() == numInputs);
+  return inputTensorValues;
+}
+
+static void eraseUnusedValuesFromMap(ValueToParamPairMap& valsToParamsMap) {
+  auto it = valsToParamsMap.begin();
+  while (it != valsToParamsMap.end()) {
+    if (!it->first->hasUses()) {
+      it = valsToParamsMap.erase(it);
+    } 
+    else {
+      ++it;
+    }
+  }
+}
+
+// This method updates the block in-place to fold all the one-time 
+// constant-based computations/ops into an initializer node.
+void ConstantFoldONNX(Block* b, ParamMap& paramsDict) {
+  auto sourceNode = b->param_node();
```

Comment:
These changes look good! My one remaining nit is that we shouldn't operate on an entity called `sourceNode`, since it more clearly expresses your intent to operate on block inputs (as a reminder, `block->inputs()` is equivalent to `param_node()->outputs()`).

Reply:
Good point. Updated.